src/kafka/ztf.rs

```diff
@@ -74,16 +75,38 @@ impl AlertConsumer for ZtfAlertConsumer {
 
     #[instrument(skip(self))]
     async fn consume(&self, timestamp: i64) {
-        let partitions_per_thread = 15 / self.n_threads;
-        let mut partitions = vec![vec![]; self.n_threads];
-        for i in 0..15 {
-            partitions[i / partitions_per_thread].push(i as i32);
-        }
-
         // ZTF uses nightly topics, and no user/pass (IP whitelisting)
         let date = chrono::DateTime::from_timestamp(timestamp, 0).unwrap();
         let topic = format!("ztf_{}_programid{}", date.format("%Y%m%d"), self.program_id);
 
+        // call check_kafka_topic_partitions to ensure the topic exists (it returns the number of partitions)
+        // do so in a while until the topic exists
+        let mut nb_partitions = 0;
+        let mut topic_exists = false;
+        while !topic_exists {
+            match check_kafka_topic_partitions(ZTF_SERVER_URL, &topic).await {
+                Ok(Some(partitions)) => {
+                    nb_partitions = partitions;
+                    topic_exists = true;
+                }
+                Ok(None) => {
+                    error!("Topic {} does not exist yet, retrying...", topic);
+                    std::thread::sleep(core::time::Duration::from_secs(5));
+                }
+                Err(e) => {
+                    error!("Error checking topic {}: {}", topic, e);
+                    return;
+                }
+            }
+        }
+
+        let n_threads = self.n_threads.clone().min(nb_partitions);
```

Comment:
I *think* this shouldn't be necessary. If there are fewer partitions than threads, I think kafka just doesn't assign any partitions to the leftover consumers.

That said, I'm totally ok with conserving threads as you're doing here.

Reply:
I'd rather prevent spawning threads if those will be idle anyways